### PR TITLE
Added '@typescript-eslint/recommended' to ESLint

### LINKED ISSUE #5541: Add new eslint rule typescript-plugin/recommended
Broken out of #5531
Requires #5540

Add `typescript-plugin/recommended` to our eslint configuration as an override applied to all `.tsx` files
Resolve all lint issues

### PR DESCRIPTION
**Pull request by omlondhe**
Fixes #5541

## Description
Added `typescript-plugin/recommended` to the `.eslintrc.json` file for linting purpose.

## Validation
I added rules to ESLint file and thus, when a developer tries to write a code not following the guidelines, then warnings are shown. Same warnings are shown while running `npx turbo lint`.

## Related Issues
Fixes #5541 raised by @bmuenzenmeyer 

### Check List
- [x] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [x] I have run `npx turbo lint` to ensure the code follows the style guide. And run `npx turbo lint:fix` to fix the style errors if necessary.
- [x] I have run `npx turbo format` to ensure the code follows the style guide.
- [ ] I have run `npx turbo test` to check if all tests are passing, and/or `npx turbo test:snapshot` to update snapshots if I created and/or updated React Components.
- [ ] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -4,15 +4,27 @@
     "next",
     "next/core-web-vitals",
     "prettier",
-    "plugin:storybook/recommended"
+    "plugin:storybook/recommended",
+    "plugin:@typescript-eslint/recommended"
```
Comment (ovflowd) [inline .eslintrc.json]: This extend should not be here. It should be below line 20.

File: .eslintrc.json [hunk 2]
```diff
@@ -4,15 +4,27 @@
...
   ],
+  "parser": "@typescript-eslint/parser",
   "overrides": [
+    {
```
Comment (ovflowd) [inline .eslintrc.json:10]: What is this rule for?

File: .eslintrc.json [hunk 3]
```diff
@@ -4,15 +4,27 @@
...
+      "files": ["**/*.mjs"],
+      "rules": {
+        "@typescript-eslint/no-empty-function": 1
+      }
+    },
     {
       "files": ["**/*.ts?(x)"],
       "plugins": ["@typescript-eslint", "no-relative-import-paths"],
       "rules": {
         "@typescript-eslint/consistent-type-imports": "error",
```
Comment (ovflowd) [inline .eslintrc.json:16]: I assume this rule and the one from the next line are already part of the recommended rules?

File: .eslintrc.json [hunk 4]
```diff
@@ -4,15 +4,27 @@
...
         "@typescript-eslint/no-unused-vars": "error",
+        "@typescript-eslint/no-non-null-assertion": 1,
```
Comment (ovflowd) [inline .eslintrc.json]: Could you comment on why you added these rules? 🤔

File: .eslintrc.json (hunk 1 above)
Comment (omlondhe) [reply .eslintrc.json]: Okay, got it!
Moved it there.

File: .eslintrc.json (hunk 2 above)
Comment (omlondhe) [reply .eslintrc.json:10]: It was for `.mjs` files, removed it as I have moved `"plugin:@typescript-eslint/recommended"` line to the line no. 20

File: .eslintrc.json (hunk 3 above)
Comment (omlondhe) [reply .eslintrc.json:16]: This rule is now removed.

Comment (omlondhe) [reply .eslintrc.json:16]: Yes it is part of recommended rules.

File: .eslintrc.json (hunk 4 above)
Comment (omlondhe) [reply .eslintrc.json]: These rules are present in the `"plugin:@typescript-eslint/recommended"` but they are specified with value `2`, thus they throw linting error when linter executes, I overwrote the value to be `1` to change the `error` property to `warning`.

May I leave it like that or change it to `2`/remove it?

Comment (ovflowd) [reply .eslintrc.json]: Why would you change the default behaviour? If they throw errors, they should throw errors 😅 

And then the errors should be fixed...

File: .eslintrc.json (hunk 2 above)
Comment (ovflowd) [reply .eslintrc.json:10]: Gotcha!

File: .eslintrc.json (hunk 4 above)
Comment (omlondhe) [reply .eslintrc.json]: Yes, right.
But existing code is getting affected now, may I still leave the default behaviour?

Comment (ovflowd) [reply .eslintrc.json]: In other words, if ESlint starts to give errors, then we should fix the Errors 👀 if you need guidance, just let us know :)

Comment (omlondhe) [reply .eslintrc.json]: Screenshot 2023-07-28 at 15 50 26

See this, these are the errors!

Comment (omlondhe) [reply .eslintrc.json]: May I go ahead and fix these?

Comment (omlondhe) [reply .eslintrc.json]: Pushed new changes, can you please review them?

Comment (ovflowd) [reply .eslintrc.json]: Will do :)

Comment (omlondhe) [reply .eslintrc.json]: Okay, thank you!
Tell me if there's any change needed, will make that change!

File: .eslintrc.json [hunk 5]
```diff
@@ -10,6 +10,8 @@
     {
       "files": ["**/*.ts?(x)"],
       "plugins": ["@typescript-eslint", "no-relative-import-paths"],
+      "extends": ["plugin:@typescript-eslint/recommended"],
+      "parser": "@typescript-eslint/parser",
       "rules": {
         "@typescript-eslint/consistent-type-imports": "error",
         "@typescript-eslint/no-unused-vars": "error",
```
Comment (bmuenzenmeyer) [inline .eslintrc.json]: as far as I can tell, this rule is included within @typescript-eslint/recommend and should not need to be redeclared here

https://typescript-eslint.io/rules/no-unused-vars

does the lint result change without this line?

```suggestion
```

Comment (omlondhe) [reply .eslintrc.json]: No, lint results does not change with this change.
Committing this change, thank you for the sugestion.

Comment (omlondhe) [reply .eslintrc.json]: I committed the change, it did not result  in any error on local machine but it resulted in some errors in pre-checks. Thus reverted back the change.

**Comment by ovflowd** [comment]
(Sorry for spamming your PR, attempting to fix some GitHub Actions stuff, and your PR became the guinea pig PR 🙈)

**Comment by omlondhe** [comment]
> (Sorry for spamming your PR, attempting to fix some GitHub Actions stuff, and your PR became the guinea pig PR 🙈)

I did not understand what you actually said, can you explain please?

**Comment by ovflowd** [comment]
> > (Sorry for spamming your PR, attempting to fix some GitHub Actions stuff, and your PR became the guinea pig PR 🙈)
> 
> I did not understand what you actually said, can you explain please?

I mentioned that I'm updating a few GitHub Workflows, and because of that, your PR is being spammed with a few comments from Vercel, for example.

**Comment by omlondhe** [comment]
> > > (Sorry for spamming your PR, attempting to fix some GitHub Actions stuff, and your PR became the guinea pig PR 🙈)
> > 
> > 
> > I did not understand what you actually said, can you explain please?
> 
> I mentioned that I'm updating a few GitHub Workflows, and because of that, your PR is being spammed with a few comments from Vercel, for example.



> > > (Sorry for spamming your PR, attempting to fix some GitHub Actions stuff, and your PR became the guinea pig PR 🙈)
> > 
> > 
> > I did not understand what you actually said, can you explain please?
> 
> I mentioned that I'm updating a few GitHub Workflows, and because of that, your PR is being spammed with a few comments from Vercel, for example.

Oh! I see! No worries, making the platform great is what’s important! I’ll bear with it!

**Comment by ovflowd** [comment]
> > > > (Sorry for spamming your PR, attempting to fix some GitHub Actions stuff, and your PR became the guinea pig PR 🙈)
> > > 
> > > 
> > > I did not understand what you actually said, can you explain please?
> > 
> > 
> > I mentioned that I'm updating a few GitHub Workflows, and because of that, your PR is being spammed with a few comments from Vercel, for example.
> 
> > > > (Sorry for spamming your PR, attempting to fix some GitHub Actions stuff, and your PR became the guinea pig PR 🙈)
> > > 
> > > 
> > > I did not understand what you actually said, can you explain please?
> > 
> > 
> > I mentioned that I'm updating a few GitHub Workflows, and because of that, your PR is being spammed with a few comments from Vercel, for example.
> 
> Oh! I see! No worries, making the platform great is what’s important! I’ll bear with it!

Thanks, I appreciate it! It's been a little bit rough since testing these GitHub Actions changes locally is rough

**Comment by ovflowd** [comment]
FYI @omlondhe let me know once you've updated this PR? I still see ESlint giving warnings about some TypeScript issues?

**Comment by omlondhe** [comment]
> > > > > (Sorry for spamming your PR, attempting to fix some GitHub Actions stuff, and your PR became the guinea pig PR 🙈)
> > > > 
> > > > 
> > > > I did not understand what you actually said, can you explain please?
> > > 
> > > 
> > > I mentioned that I'm updating a few GitHub Workflows, and because of that, your PR is being spammed with a few comments from Vercel, for example.
> > 
> > 
> > > > > (Sorry for spamming your PR, attempting to fix some GitHub Actions stuff, and your PR became the guinea pig PR 🙈)
> > > > 
> > > > 
> > > > I did not understand what you actually said, can you explain please?
> > > 
> > > 
> > > I mentioned that I'm updating a few GitHub Workflows, and because of that, your PR is being spammed with a few comments from Vercel, for example.
> > 
> > 
> > Oh! I see! No worries, making the platform great is what’s important! I’ll bear with it!
> 
> Thanks, I appreciate it! It's been a little bit rough since testing these GitHub Actions changes locally is rough

Yes! I was learning this workflow thing I understand!

**Comment by omlondhe** [comment]
> FYI @omlondhe let me know once you've updated this PR? I still see ESlint giving warnings about some TypeScript issues?

Yeah updated the PR and tried to solve some errors too! There are still 6 warnings, I have resolve them too?

**Comment by ovflowd** [comment]
Since those warnings got introduced by the eslint recommended plugin, yes 🙃

If you need help let me know!

**Comment by omlondhe** [comment]
> Since those warnings got introduced by the eslint recommended plugin, yes 🙃
> 
> If you need help let me know!

Okay, I’ll do that! Will ask you if I need any help!

**Comment by omlondhe** [comment]
> Since those warnings got introduced by the eslint recommended plugin, yes 🙃
> 
> If you need help let me know!

Tried to solve those warnings, but am unable to solve 3 of them.
image
Can you please tell me how can I solve these?

**Comment by bmuenzenmeyer** [comment]
@omlondhe 

for `AnchoredHeading` you can do this

```tsx
const HeadingLevelTag = `h${level}` as unknown as FC<React.PropsWithChildren>
```

for `getNodejsChangelog` I opted to ignore the rule at the file level with a comment as to why

```
// ignoring concern for missing versions here - as the data flow from Node.js releases is rigid and we can make assumptions here
/* eslint-disable @typescript-eslint/no-non-null-assertion */
```

**Comment by omlondhe** [comment]
> @omlondhe
> 
> for `AnchoredHeading` you can do this
> 
> ```tsx
> const HeadingLevelTag = `h${level}` as unknown as FC<React.PropsWithChildren>
> ```
> 
> for `getNodejsChangelog` I opted to ignore the rule at the file level with a comment as to why
> 
> ```
> // ignoring concern for missing versions here - as the data flow from Node.js releases is rigid and we can make assumptions here
> /* eslint-disable @typescript-eslint/no-non-null-assertion */
> ```

Thank you for the help @bmuenzenmeyer !
Done with the changes, this PR is updated now and there is no any error or warning.

Can you please go through it and add your feedback?

**Review by bmuenzenmeyer — APPROVED** [review]
nice work!

**Comment by omlondhe** [comment]
> nice work!

Thank you @bmuenzenmeyer , I would love to contribute more!

**Pull request merged**